File: mcp-server-go/internal/tools/status.go
// Status tool — 1 tool for project status aggregation.
// Port of mcp-server/src/tools/status.ts
package tools

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/mark3labs/mcp-go/mcp"

	"github.com/Chachamaru127/harness-mem/mcp-server-go/internal/types"
	"github.com/Chachamaru127/harness-mem/mcp-server-go/internal/util"
)

var (
	reTODO = regexp.MustCompile(`cc:TODO`)
	reWIP  = regexp.MustCompile(`cc:WIP`)
	reDONE = regexp.MustCompile(`cc:DONE`)
)

var ssotFiles = []string{
	".claude/memory/decisions.md",
	".claude/memory/patterns.md",
	"AGENTS.md",
	"CLAUDE.md",
}

func StatusToolDefs() []ToolDef {
	return []ToolDef{
		{statusTool, func(_ context.Context, args map[string]any) types.ToolResult { return handleStatus(args) }},
	}
}

var statusTool = mcp.NewTool("harness_status",
	mcp.WithDescription("Get current project status including Plans.md progress, active sessions, and recent activity"),
	mcp.WithBoolean("verbose", mcp.Description("Include detailed information")),
)

func handleStatus(args map[string]any) types.ToolResult {
	verbose := argBool(args, "verbose", false)
	projectRoot := util.GetProjectRoot()

	var sb strings.Builder
	sb.WriteString("**Harness Status**\n\n")
	sb.WriteString(fmt.Sprintf("Project: %s\n", filepath.Base(projectRoot)))

	// Version
	versionFile := filepath.Join(projectRoot, ".claude-code-harness-version")
	if data, err := os.ReadFile(versionFile); err == nil {
		sb.WriteString(fmt.Sprintf("Harness: v%s\n", strings.TrimSpace(string(data))))
	}
	sb.WriteString("\n")

	// Plans status
	plansPath := filepath.Join(projectRoot, "Plans.md")
	if data, err := os.ReadFile(plansPath); err == nil {
		content := string(data)
		todo := len(reTODO.FindAllString(content, -1))
		wip := len(reWIP.FindAllString(content, -1))
		done := len(reDONE.FindAllString(content, -1))
		total := todo + wip + done
		progress := 0
		if total > 0 {
			progress = done * 100 / total
		}
		sb.WriteString("**Plans.md**\n")
		sb.WriteString(fmt.Sprintf("  TODO: %d\n  WIP: %d\n  Done: %d\n  Progress: %d%%\n\n", todo, wip, done, progress))
	} else {
		sb.WriteString("Plans.md: Not found\n\n")
	}

	// Sessions
	sessionCount := getActiveSessionCount()
	unreadCount := getUnreadMessageCount(projectRoot)
	sb.WriteString("**Sessions**\n")
	sb.WriteString(fmt.Sprintf("  Active: %d\n  Unread messages: %d\n\n", sessionCount, unreadCount))

	// Verbose
	if verbose {
		sb.WriteString(fmt.Sprintf("**Project Root**: %s\n\n**SSOT Files**:\n", projectRoot))
		for _, f := range ssotFiles {
			exists := "x"
			if _, err := os.Stat(filepath.Join(projectRoot, f)); err == nil {
				exists = "ok"
			}
			sb.WriteString(fmt.Sprintf("  [%s] %s\n", exists, f))
		}
		sb.WriteString("\n")
	}

	// Suggestion
	sb.WriteString("**Suggested Action**: ")
	if _, err := os.Stat(plansPath); os.IsNotExist(err) {
		sb.WriteString("Use harness_workflow_plan to create a plan")
	} else if data, err := os.ReadFile(plansPath); err == nil {
		content := string(data)
		todo := len(reTODO.FindAllString(content, -1))
		wip := len(reWIP.FindAllString(content, -1))
		if todo > 0 {
			sb.WriteString(fmt.Sprintf("Use harness_workflow_work to implement %d pending task(s)", todo))
		} else if wip > 0 {
			sb.WriteString(fmt.Sprintf("Continue working on %d in-progress task(s)", wip))
		} else {
			sb.WriteString("All tasks complete! Use harness_workflow_review to review changes")
		}
	}

	return types.SuccessText(sb.String())
}

func getActiveSessionCount() int {
	sessions := util.SafeReadJSON[map[string]Session](
		filepath.Join(util.GetProjectRoot(), util.ActiveSessionsFile),
		map[string]Session{},
	)
	now := float64(time.Now().Unix())
	count := 0
	for _, s := range sessions {
		if now-s.LastSeen < float64(util.StaleThresholdSeconds) {
			count++
		}
	}
	return count
}

func getUnreadMessageCount(projectRoot string) int {
	// Broadcast is markdown, not JSON — just count recent entries
	data, err := os.ReadFile(filepath.Join(projectRoot, util.BroadcastFile))
	if err != nil {
		return 0
	}
	cutoff := time.Now().Add(-1 * time.Hour)
	matches := broadcastRegex.FindAllStringSubmatch(string(data), -1)
	count := 0
	for _, m := range matches {
		t, err := time.Parse("2006-01-02T15:04:05Z", m[1])
		if err == nil && t.After(cutoff) {
			count++
		}
	}
	return count
}

File: mcp-server/src/tools/status.ts
/**
 * Status Tools
 *
 * Project status and synchronization tools.
 */

import { type Tool } from "@modelcontextprotocol/sdk/types.js";
import * as fs from "fs";
import * as path from "path";
import {
  ACTIVE_SESSIONS_FILE,
  BROADCAST_FILE,
  STALE_THRESHOLD_SECONDS,
  type PlansScopeArgs,
  resolvePlansTarget,
  safeReadJSON,
} from "../utils.js";

const PLANS_SCOPE_SCHEMA = {
  cwd: {
    type: "string",
    description:
      "Caller working directory used to locate the client project's Plans.md. Required unless project or plans_path is supplied.",
  },
  project: {
    type: "string",
    description:
      "Absolute filesystem project path for Plans.md operations. Required unless cwd or plans_path is supplied; short project keys are not accepted.",
  },
  plans_path: {
    type: "string",
    description:
      "Absolute path to a Plans.md file. Required unless cwd or project is supplied; must point to Plans.md.",
  },
} as const;

// Types for session and message data
interface SessionData {
  lastSeen: number;
  [key: string]: unknown;
}

interface BroadcastMessage {
  timestamp: string;
  [key: string]: unknown;
}

/** Default message window: 1 hour in milliseconds */
const MESSAGE_WINDOW_MS = 3600000;
const DONE_MARKER_PATTERN = /cc:(?:完了|[dD][oO][nN][eE])(?:\s|$|\[|\(|<|\|)/g;

// Tool definitions
export const statusTools: Tool[] = [
  {
    name: "harness_status",
    description:
      "Get current project status including Plans.md progress, active sessions, and recent activity",
    inputSchema: {
      type: "object",
      properties: {
        verbose: {
          type: "boolean",
          description: "Include detailed information",
        },
        ...PLANS_SCOPE_SCHEMA,
      },
      required: [],
    },
  },
];

// Helper functions using shared utilities
function getPlansStatus(plansPath: string): { todo: number; wip: number; done: number } | null {
  if (!fs.existsSync(plansPath)) {
    return null;
  }

  const content = fs.readFileSync(plansPath, "utf-8");
  return {
    todo: (content.match(/cc:TODO/g) || []).length,
    wip: (content.match(/cc:WIP/g) || []).length,
    done: (content.match(DONE_MARKER_PATTERN) || []).length,
  };
}

function getSessionCount(projectRoot: string): number {
  const sessions = safeReadJSON<Record<string, SessionData>>(
    path.join(projectRoot, ACTIVE_SESSIONS_FILE),
    {}
  );
  const now = Date.now() / 1000;

  return Object.values(sessions).filter(
    (s) => now - s.lastSeen < STALE_THRESHOLD_SECONDS
  ).length;
}

function getUnreadMessageCount(projectRoot: string): number {
  const messages = safeReadJSON<BroadcastMessage[]>(path.join(projectRoot, BROADCAST_FILE), []);
  const cutoff = Date.now() - MESSAGE_WINDOW_MS;

  return messages.filter(
    (m) => new Date(m.timestamp).getTime() > cutoff
  ).length;
}

function getHarnessVersion(projectRoot: string): string | null {
  const versionFile = path.join(projectRoot, ".claude-code-harness-version");
  if (fs.existsSync(versionFile)) {
    return fs.readFileSync(versionFile, "utf-8").trim();
  }
  return null;
}

// SSOT files to check for project health
const SSOT_FILES = [
  ".claude/memory/decisions.md",
  ".claude/memory/patterns.md",
  "AGENTS.md",
  "CLAUDE.md",
] as const;

// Tool handlers
export async function handleStatusTool(
  name: string,
  args: Record<string, unknown> | undefined
): Promise<{ content: Array<{ type: string; text: string }>; isError?: boolean }> {
  switch (name) {
    case "harness_status":
      return handleStatus(args as ({ verbose?: boolean } & PlansScopeArgs) | undefined);

    default:
      return {
        content: [{ type: "text", text: `Unknown status tool: ${name}` }],
        isError: true,
      };
  }
}

function handleStatus(args: ({ verbose?: boolean } & PlansScopeArgs) | undefined): {
  content: Array<{ type: string; text: string }>;
  isError?: boolean;
} {
  const { verbose = false } = args ?? {};

  const resolved = resolvePlansTarget(args);
  if (!resolved.ok) {
    return {
      content: [{ type: "text", text: resolved.message }],
      isError: true,
    };
  }

  const { projectRoot, plansPath } = resolved.target;
  const plansStatus = getPlansStatus(plansPath);
  const sessionCount = getSessionCount(projectRoot);
  const unreadCount = getUnreadMessageCount(projectRoot);
  const harnessVersion = getHarnessVersion(projectRoot);

  let status = `📊 **Harness Status**\n\n`;

  // Project info
  status += `📁 Project: ${path.basename(projectRoot)}\n`;
  if (harnessVersion) {
    status += `🔧 Harness: v${harnessVersion}\n`;
  }
  status += `\n`;

  // Plans status
  if (plansStatus) {
    const total = plansStatus.todo + plansStatus.wip + plansStatus.done;
    const progress =
      total > 0 ? Math.round((plansStatus.done / total) * 100) : 0;

    status += `📋 **Plans.md**\n`;
    status += `├─ TODO: ${plansStatus.todo}\n`;
    status += `├─ WIP: ${plansStatus.wip}\n`;
    status += `├─ Done: ${plansStatus.done}\n`;
    status += `└─ Progress: ${progress}%\n\n`;
  } else {
    status += `📋 Plans.md: Not found\n\n`;
  }

  // Session info
  status += `👥 **Sessions**\n`;
  status += `├─ Active: ${sessionCount}\n`;
  status += `└─ Unread messages: ${unreadCount}\n\n`;

  // Verbose info
  if (verbose) {
    status += `📍 **Project Root**: ${projectRoot}\n`;

    status += `\n📄 **SSOT Files**:\n`;
    for (const file of SSOT_FILES) {
      const exists = fs.existsSync(path.join(projectRoot, file));
      status += `${exists ? "✅" : "❌"} ${file}\n`;
    }
  }

  // Next action suggestion
  status += `\n💡 **Suggested Action**: `;
  if (!plansStatus) {
    status += `Use harness_workflow_plan with the same cwd/project/plans_path to create a plan`;
  } else if (plansStatus.todo > 0) {
    status += `Use harness_workflow_work with the same cwd/project/plans_path to implement ${plansStatus.todo} pending task(s)`;
  } else if (plansStatus.wip > 0) {
    status += `Continue working on ${plansStatus.wip} in-progress task(s)`;
  } else {
    status += `All tasks complete! Use harness_workflow_review to review changes`;
  }

  return {
    content: [{ type: "text", text: status }],
  };
}
